frontend-wasm/src/code_translator/mod.rs

```diff
@@ -191,7 +192,10 @@ pub fn translate_operator(
         }
         Operator::I64ExtendI32U => {
             let val = state.pop1();
-            state.push1(builder.ins().zext(val, I64, span));
+            let u32_val = builder.ins().cast(val, U32, span);
```

Comment:
I really think we're going to need to reconsider the impedance mismatch between how Wasm represents integral types and how the IR does it, or at least provide an additional set of APIs that are more akin to `transmute` than conversion. Right now, the code we emit when casting signed -> unsigned, or larger bitwidth to smaller must check that the value fits in the target type. However, due to the way Wasm represents things, this results in a lot of redundant code that simply doesn't need to exist, and it isn't code we can easily remove later (we might be able to do some of it with peephole optimization, but it will be fragile to the distance between the conversion in, and the conversion out of a given type).

We should probably discuss this on a call, because it'll be a lot to cover, but here's some ideas we could explore:

1. Modify the IR type system to be closer to Wasm, this has some downsides for non-Wasm frontends, but would at least make the code we generate from Wasm a lot more optimal
2. Modify the IR type system to have signless integral types, and/or provide a set of instructions that interpret the value of an integral type based on whether the instruction is a signed or unsigned variant (e.g. in addition to `add` which operates on `Type::I32` or `Type::U32` according to the type, we'd also have `add_s` and `add_u` which treat the bits of a `Type::I32` or `Type::U32` as signed or unsigned depending on the instruction rather than the value type). The type returned from such an instruction would always be the same as the input type.
3. Rather than changing the type system or instruction set, we would add a new instruction for transmuting between types. Transmutation would perform no range checks, and would do the bare minimum necessary to treat a value of a given source type as the desired target type. For example, `transmute i32 : u32` would result in no code being emitted during codegen.

Personally, I'm leaning towards 2. While 1 would be an option, it's a more significant change, and weakens the type system, which might be fine, but is a tradeoff. 3 feels like an incomplete hack really, but is probably the simplest option to help with the poor codegen output. The problem is that sign/zero-extension is also affected by this, so we'd either need to weaken the guarantees of those instructions, or make a theoretical `transmute` handle extension as well, so you could do things like transmute between `i32` and `i64`.

Thoughts?

Reply:
My least favorite option is 1. I find the second option quite confusing and somewhat close to the first option, so I'm leaning toward the third option for now. Let's discuss it on a call.